File: leetcode2/2medium/김대원/Q331.java
// 1. 아이디어 : 노드가 존재할 때, 자식 노드 개수 판단
// 2. 시간복잡도 : O(N)
// 3. 자료구조/알고리즘 : 그리디

class Solution {

    // 9: cnt = --: 0 -> 2
    // 3: cnt = --: 1 -> 3
    // 4: cnt = --: 2 -> 4
    // 1: cnt = --: 1 -> 3
    // 2: cnt = --: 0 -> 2
    // 6: cnt = --; 0 -> 2

    public boolean isValidSerialization(String preorder) {
        int child = 1;
        String[] strs = preorder.split(",");
        for (String str : strs) {
            child--;
            if (child >= 0)  {
                if (!str.equals("#")) child += 2;
            } else break;
        }

        return child == 0;
    }
}

File: leetcode2/2medium/강병호/Q331.py
import string

class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        preorder_list = list(map(str, preorder.split(",")))
        print(preorder_list)

        node_count = 1

        # 숫자면 최대 2개의 노드 생성 가능
        for data in preorder_list:
            node_count -= 1 # null이든 숫자든 하나 사용

            if node_count < 0:
                return False

            if data != '#': # 숫자면 2개 노드 생성가능
                node_count += 2
        
        if node_count == 0:
            return True
        else:
            return False

            




if __name__ == "__main__":
    sol = Solution()
    sol.isValidSerialization("9,3,4,#,#,1,#,#,2,#,6,#,#")